File: day11/AOC.py
import re
from time import perf_counter

# Password requirements:
# 1. Must include one increasing straight of at least three letters
# 2. Must not contain the letters i, o, or l
# 3. Must contain at least two different, non-overlapping pairs of letters
def password_validity(password):
    # This regex is very long and ugly, but it works
    if re.search(r"abc|bcd|cde|def|efg|fgh|ghi|hij|ijk|jkl|klm|lmn|mno|nop|opq|pqr|qrs|rst|stu|tuv|uvw|vwx|wxy|xyz", password) == None:
        return False
    if re.search(r"[iol]", password) != None:
        return False
    if re.search(r"(.)\1.*(.)\2", password) == None:
        return False
    return True

# Slow, but gets the job done. I takes ~ 9-10 seconds to run the tests
# Can be optimized by skipping paasswords that contain i, o, or l
# Implemented it, speed up about 2 seconds on the tests, no change real input
def next_password(password):
    p = list(password)
    if p[-1] == "z":
        p[-1] = "a"
        index = -2
        while p[index] == "z":
            p[index] = "a"
            index -= 1
        p[index] = chr(ord(p[index]) + 1)
    else:
        p[-1] = chr(ord(p[-1]) + 1)
    password = "".join(p)
    return password

def part1(input):
    password = next_password(input)
    while not password_validity(password):
        password = next_password(password)
    return password

def part2(input):
    return part1(input)

def main():
    input = open("input.txt", "r").read().splitlines()
    part1_sum = part1(input[0])
    print(f"Part 1: {part1_sum}")
    
    t1 = perf_counter()
    part2_sum = part2(part1_sum)
    t2 = perf_counter()
    print(f"Part 2: {part2_sum}")
    print(f"It part 2 took {t2 - t1} seconds") # ~ 1 second, tests take ~ 9-10 seconds

if __name__ == "__main__":
    main()

File: Go/day11/AOC.go
package main

import (
	"fmt"
	"os"
	"strings"
)

// Password requirements:
// - It must contain at least one increasing straight of three letters. (e.g. abc, bcd, cde, etc.)
// - It cannot contain the letters i, o, or l.
// - It must contain at least two different, non-overlapping pairs of letters. (e.g. aa, bb, or zz)
func passwordValidator(password string) bool {
	// Check for increasing straight of three letters
	increasing := false
	for i := 0; i < len(password)-2; i++ {
		if password[i]+1 == password[i+1] && password[i]+2 == password[i+2] {
			increasing = true
			break
		}
	}

	if !increasing {
		return false
	}

	// Check for i, o, or l
	if strings.Contains(password, "i") || strings.Contains(password, "o") || strings.Contains(password, "l") {
		return false
	}

	// Check for at least two different, non-overlapping pairs of letters
	pairs := 0
	for i := 0; i < len(password)-1; i++ {
		if password[i] == password[i+1] {
			pairs++
			i++
		}
	}
	return pairs >= 2
}

func nextPassword(password string) string {
	p := []byte(password)
	if p[len(p)-1] == 'z' {
		p[len(p)-1] = 'a'
		index := len(p) - 2
		for index >= 0 && p[index] == 'z' {
			p[index] = 'a'
			index--
		}
		if index >= 0 {
			p[index]++
		} else {
			p = append([]byte{'a'}, p...)
		}
	} else {
		p[len(p)-1]++
	}
	return string(p)
}

func nextValidPassword(password string) string {
	for {
		// fmt.Printf("Checking %s\n", password)
		if password == "ghjaabcc" {
			fmt.Print("Found it\n")
		}
		password = nextPassword(password)
		if passwordValidator(password) {
			return password
		}
	}
}

func part1(input string) string {
	return nextValidPassword(input)
}

func part2(input string) string {
	return nextValidPassword(input)
}

func main() {
	input, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Print("Error reading file\n")
		panic(err)
	}

	lines := strings.Split(string(input), "\n")
	part1_ans := part1(lines[0])
	fmt.Printf("Part 1: %s\n", part1_ans)

	part2_ans := part2(part1_ans)
	fmt.Printf("Part 2: %s\n", part2_ans)
}
